compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -2177,14 +2178,44 @@ public void visit(BLangExpressionStmt exprStmtNode) {
         // Creates a new environment here.
         SymbolEnv stmtEnv = new SymbolEnv(exprStmtNode, this.env.scope);
         this.env.copyTo(stmtEnv);
-        BType bType = typeChecker.checkExpr(exprStmtNode.expr, stmtEnv, symTable.noType);
-        if (bType != symTable.nilType && bType != symTable.neverType && bType != symTable.semanticError &&
-                exprStmtNode.expr.getKind() != NodeKind.FAIL) {
+        typeChecker.checkExpr(exprStmtNode.expr, stmtEnv, symTable.noType);
+
+        NodeKind exprKind = exprStmtNode.expr.getKind();
+        boolean isChecked = exprKind == NodeKind.CHECK_EXPR || exprKind == NodeKind.CHECK_PANIC_EXPR;
+        BType expExprType = getExpectedTypeOfExpr(exprStmtNode.expr, isChecked);
+
+        boolean returnsNilOrNever = expExprType == symTable.nilType || expExprType == symTable.neverType
+                || (isChecked && types.isOptionalErrorType(expExprType));
+
+        if (!returnsNilOrNever && exprStmtNode.expr.getKind() != NodeKind.FAIL) {
             dlog.error(exprStmtNode.pos, DiagnosticCode.ASSIGNMENT_REQUIRED);
         }
         validateWorkerAnnAttachments(exprStmtNode.expr);
     }
 
+    // This method checks and returns the actual return type of a function/method as specified in its signature.
+    // Cannot just simply use the type of the expr for this since it's the evaluated type for the particular
+    // invocation of the function. If there was an error in evaluating the type, the type of the invocation would be
+    // semantic error.
+    private BType getExpectedTypeOfExpr(final BLangExpression expr, boolean isCheckedExpr) {
+        NodeKind exprKind = expr.getKind();
+        BLangExpression tempExpr = expr;
+
+        if (isCheckedExpr) {
+            tempExpr = ((BLangCheckedExpr) expr).expr;
+            exprKind = expr.getKind();
+        }
+
+        if (exprKind == NodeKind.INVOCATION) {
+            BInvokableSymbol symbol = (BInvokableSymbol) ((BLangInvocation) tempExpr).symbol;
+            // `symbol` is null for invocations of undefined functions. Consider the return type of undefined
+            // functions as `()`.
+            return symbol != null ? symbol.retType : symTable.nilType;
```

Comment:
Instead of returning nil, what if we return semanticError, and then avoid logging the "ASSIGNMENT_REQUIRED" error for semanticError too? That way, if by any chance we get there for non-invocation expressions with semantic error as the type, it would also be handled?

Reply:
We can't do that right? If it's not an effectively nil returning function/method call expression, we have log the assignment required error, irrespective of the type of the expression.

Reply:
What about something like `check x;`? If it is a semantic error, I don't think the "assignment required" error adds much value, and there will already be an error for `x` anyway?

Reply:
If we don't log it, once we fix the error in `x`, we'll again get the assignment required error the next time we try to run the program, which happened in the reported issue (#24839). Skipping semantic errors was what was done previously as well. The downside to it is you don't get all the errors in one go.

Reply:
But in this case, once we fix `x`, it may or may not result in the "assignment required" error, right? I'm all for showing as many errors as possible, but I'm just wondering if logging errors based on possible paths could lead to unnecessary noise? We should of course set the type properly on error.

Also, if we are going to log the "assignment required" error for such scenarios, shouldn't we also log the error when it is an undefined function?

Reply:
The general case is that the assignment is required right? Considering that this is a case where the user has attempted to use an expression as a statement. Assignments are not required only for specific cases. For example,
```ballerina
public function main() {
    int x = 10;
    checkpanic foo() + x;
}

function foo() returns error? => error("Invalid Value");
```
In the above you won't need the assignment. You just have to fix the error in the expr of the checkpanic expr.

Reply:
For undefined functions it's not getting logged because here, the return type of undefined functions is taken as `()`. That was done just for simplicity, since I felt there's no point going further than saying that the function is not defined.

Reply:
> The general case is that the assignment is required right? Considering that this is a case where the user has attempted to use an expression as a statement. Assignments are not required only for specific cases. For example,
> 
> ```ballerina
> public function main() {
>     int x = 10;
>     checkpanic foo() + x;
> }
> 
> function foo() returns error? => error("Invalid Value");
> ```
> 
> In the above you won't need the assignment. You just have to fix the error in the expr of the checkpanic expr.

I also think the more common case would be to require the assignment, but do we have to make that assumption? I'm not sure I got the example, here the error would be that you can't add nil and int right? And if the user changes `foo` to return `int|error` an assignment will be required, but if they remove `+ x` no assignment is required?



> For undefined functions it's not getting logged because here, the return type of undefined functions is taken as `()`. That was done just for simplicity, since I felt there's no point going further than saying that the function is not defined.

This was the point I was trying to make re: any expression too. If we've already logged appropriate errors, it feels like there's no point giving an error adjusting for code that will definitely have to be changed? What I'm suggesting is to kinda not special case the undefined function scenario, but do the same for all scenarios where the checking-expr's expr itself has errors.

Reply:
> I also think the more common case would be to require the assignment, but do we have to make that assumption? I'm not sure I got the example, here the error would be that you can't add nil and int right? And if the user changes foo to return int|error an assignment will be required, but if they remove + x no assignment is required?

Sorry, that should've been `checkpanic (foo() + x);`. My point was, the type of the expr here would be semantic error. So can't really figure out anything from that to decide whether there should be an assignment or not. 

> This was the point I was trying to make re: any expression too. If we've already logged appropriate errors, it feels like there's no point giving an error adjusting for code that will definitely have to be changed? What I'm suggesting is to kinda not special case the undefined function scenario, but do the same for all scenarios where the checking-expr's expr itself has errors.

The problem with this is then we get assignment expected errors for nil returning lang lib functions as well. And the same with the example above. So special cased the call-stmt scenario. 
```ballerina
function test1(){
    int x = 0;
    x.forEach(function (int i) { count = count + i;});
}
```

Reply:
We discussed this offline, and decided to revisit this if required, once we validate the following is valid according to the spec.

```ballerina
check x;
```